### PR TITLE
[DataPipe] Skip FullSync operation when world_size == 1

### PR DESCRIPTION
**Pull request by NivekT**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #1065

Differential Revision: [D43751382](https://our.internmc.facebook.com/intern/diff/D43751382)

### CODE REVIEW

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43751382).

File: torchdata/datapipes/iter/util/distributed.py [hunk 1]
```diff
@@ -231,11 +234,13 @@ def __setstate__(self, state):
         self._done_callback = False
 
     def pause(self):
-        raise RuntimeError("`pause` is not supported for FullSync at the moment.")
+        if self._world_size > 1:
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/distributed.py:237]:
```suggestion
        if self._world_size > 1 and self._executor is not None:
```

File: torchdata/datapipes/iter/util/distributed.py [hunk 2]
```diff
@@ -231,11 +234,13 @@ def __setstate__(self, state):
...
+            raise RuntimeError("`pause` is not supported for FullSync at the moment.")
         # if self._executor is not None:
         #     self._executor.shutdown()
         #     self._executor = None
 
     def resume(self):
-        raise RuntimeError("`resume` is not supported for FullSync at the moment.")
+        if self._world_size > 1:
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/distributed.py:244]: Ditto

**Review by ejguan — APPROVED** [review]
Approve with one comment

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43751382).

**Review by ejguan — APPROVED** [review]

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43751382).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43751382).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43751382).

File: torchdata/datapipes/iter/util/distributed.py [hunk 3]
```diff
@@ -171,31 +171,34 @@ def _callback_fn(self, exp: Expected) -> None:
...
-                    lambda: self._done_callback is True,
-                    self.timeout,
-                )
-                if not is_success:
-                    raise PrefetchTimeoutError(self.timeout)
-                if self._error is not None:
-                    raise self._error
-                if bool(self._sync_counter < self._world_size):
+        if self._world_size == 1:  # The below functionalities are not needed if `_world_size == 1`
+            yield from self.datapipe
```
Comment (ejguan) [inline torchdata/datapipes/iter/util/distributed.py:182]: Nit comment and let you decide if you want to do it.
you can `return` after `yield from self.datapipe` within the clause of `if self._world_size == 1:` to reduce the `else` clause.

**Review by ejguan — APPROVED** [review]
Can we add a test?

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43751382).

**Comment by NivekT** [comment]
@nivekt has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43751382).

**Pull request closed without merging**